Index: pango/pango-utils.c
===================================================================
--- a/pango/pango-utils.c	(revision 2425)
+++ b/pango/pango-utils.c	(working copy)
@@ -567,6 +567,20 @@ read_config_file (const char *filename,
	      gchar *value =  g_key_file_get_value(key_file, group, key, &key_error);
	      if (value != NULL)
		{
+                  if (!g_path_is_absolute (value))
+                    {
+                      gchar *dirname;
+                      gchar *absolute_path;
+
+                      dirname = g_path_get_dirname (filename);
+                      absolute_path = g_build_filename (dirname, value, NULL);
+
+                      g_free (dirname);
+                      g_free (value);
+
+                      value = absolute_path;
+                    }
+
		  g_hash_table_insert (ht,
				       g_strdup_printf ("%s/%s", group, key),
				       value);
