From 3cab26a0468bab855ed2eb13e4f334176e109483 Mon Sep 17 00:00:00 2001
From: Kristian Rietveld <kris@lanedo.com>
Date: Tue, 19 Mar 2013 11:23:09 +0100
Subject: [PATCH 1/2] Simply process entire bitmap obtained from the core text
 font

---
 pango/pangocoretext.c |    3 +--
 1 files changed, 1 insertions(+), 2 deletions(-)

diff --git a/pango/pangocoretext.c b/pango/pangocoretext.c
index 36bcd82..1d46271 100644
--- a/pango/pangocoretext.c
+++ b/pango/pangocoretext.c
@@ -89,8 +89,7 @@ ct_font_descriptor_get_coverage (CTFontDescriptorRef desc)
   bitmap = CFCharacterSetCreateBitmapRepresentation (kCFAllocatorDefault,
                                                      charset);

-  /* We only handle the BMP plane */
-  length = MIN (CFDataGetLength (bitmap), 8192);
+  length = CFDataGetLength (bitmap);
   ptr = CFDataGetBytePtr (bitmap);

   /* FIXME: can and should this be done more efficiently? */
--
1.7.4.4
