From 388e70cee0897fb743531ec001c1e6f796e2c27f Mon Sep 17 00:00:00 2001
From: Michael Natterer <mitch@gimp.org>
Date: Fri, 1 Mar 2013 15:06:20 +0100
Subject: [PATCH 35/68] gtk: remove the overlay scrollbar grab on unrealize()

---
 gtk/gtkscrolledwindow.c |    6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/gtk/gtkscrolledwindow.c b/gtk/gtkscrolledwindow.c
index 4159af8..7200462 100644
--- a/gtk/gtkscrolledwindow.c
+++ b/gtk/gtkscrolledwindow.c
@@ -3256,6 +3256,12 @@ gtk_scrolled_window_unrealize (GtkWidget *widget)
   GtkScrolledWindow *scrolled_window = GTK_SCROLLED_WINDOW (widget);
   GtkScrolledWindowPrivate *priv = GTK_SCROLLED_WINDOW_GET_PRIVATE (scrolled_window);

+  if (priv->sb_pointer_grabbed)
+    {
+      gtk_grab_remove (widget);
+      priv->sb_pointer_grabbed = FALSE;
+    }
+
   gdk_window_set_user_data (priv->overshoot_window, NULL);
   gdk_window_destroy (priv->overshoot_window);
   priv->overshoot_window = NULL;
--
1.7.10.2 (Apple Git-33)
