diff --git a/gdk/quartz/gdkevents-quartz.c b/gdk/quartz/gdkevents-quartz.c
index 23e879c..b7b2665 100644
--- a/gdk/quartz/gdkevents-quartz.c
+++ b/gdk/quartz/gdkevents-quartz.c
@@ -831,7 +831,7 @@ find_window_for_ns_event (NSEvent *nsevent,
                       {
                         g_warning ("grab_toplevel == NULL");
                       }
-                    else
+                    else if (!gdk_window_is_destroyed (grab_toplevel))
                       {
                         get_window_point_from_screen_point (grab_toplevel,
                                                             screen_point, x, y);
