From 1646b8a78f00cbddc00f6138d80fbc42dec82dac Mon Sep 17 00:00:00 2001
From: iain holmes <iain@xamarin.com>
Date: Tue, 13 Sep 2016 16:06:39 +0100
Subject: [PATCH] [a11y] Make the GtkPaned widget emit GtkContainer::add
 signals

---
 gtk/gtkpaned.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/gtk/gtkpaned.c b/gtk/gtkpaned.c
index f6225f0..0e3afd6 100644
--- a/gtk/gtkpaned.c
+++ b/gtk/gtkpaned.c
@@ -1375,6 +1375,10 @@ gtk_paned_pack1 (GtkPaned  *paned,
       paned->child1_shrink = shrink;

       gtk_widget_set_parent (child, GTK_WIDGET (paned));
+
+      /* We need to emit this signal so that the accessibility system knows
+         that a widget has been added to the pane */
+      g_signal_emit_by_name (G_OBJECT (paned), "add", child);
     }
 }

@@ -1394,6 +1398,10 @@ gtk_paned_pack2 (GtkPaned  *paned,
       paned->child2_shrink = shrink;

       gtk_widget_set_parent (child, GTK_WIDGET (paned));
+
+      /* We need to emit this signal so that the accessibility system knows
+         that a widget has been added to the pane */
+      g_signal_emit_by_name (G_OBJECT (paned), "add", child);
     }
 }

@@ -1406,6 +1414,11 @@ gtk_paned_add (GtkContainer *container,

   g_return_if_fail (GTK_IS_PANED (container));

+  /* Break the add signal cycle */
+  if (widget->parent == container) {
+    return;
+  }
+
   paned = GTK_PANED (container);

   if (!paned->child1)
