From 7ad6f5dc44092c47cee1af32be0f6a061b8abb7c Mon Sep 17 00:00:00 2001
From: dseefeld <dseefeld@microsoft.com>
Date: Fri, 24 Jan 2020 19:01:58 +0000
Subject: [PATCH] Add trailing path separator to repo_root

---
 eng/common/build.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/eng/common/build.sh b/eng/common/build.sh
index 55b298f16cc..1b250d16018 100755
--- a/eng/common/build.sh
+++ b/eng/common/build.sh
@@ -202,7 +202,7 @@ function Build {
   MSBuild $_InitializeToolset \
     $bl \
     /p:Configuration=$configuration \
-    /p:RepoRoot="$repo_root" \
+    /p:RepoRoot="$repo_root/" \
     /p:Restore=$restore \
     /p:Build=$build \
     /p:Rebuild=$rebuild \
-- 
2.25.4

