From 6521f8a43f5c146efc72b72dd3b0e61344a71618 Mon Sep 17 00:00:00 2001
From: dseefeld <dseefeld@microsoft.com>
Date: Mon, 15 Mar 2021 19:00:55 +0000
Subject: [PATCH] Exclude Microsoft.Extensions.Hosting.WindowsServices from
 source-build

---
 .../Directory.Build.props                                       | 3 ++
 1 file changed, 3 insertions(+)

diff --git a/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/Directory.Build.props b/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/Directory.Build.props
index c5f4d2e6da1..5493f19bc19 100644
--- a/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/Directory.Build.props
+++ b/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/Directory.Build.props
@@ -2,5 +2,8 @@
   <Import Project="..\Directory.Build.props" />
   <PropertyGroup>
     <PackageTags>hosting</PackageTags>
+    <!-- Source build does not support Windows. -->
+    <ExcludeFromSourceBuild>true</ExcludeFromSourceBuild>
+    <IsPackable>false</IsPackable>
   </PropertyGroup>
 </Project>
-- 
2.25.4
