diff --git a/src/murrine_draw.c b/src/murrine_draw.c
index 4cab20f..06ed9b1 100644
--- a/src/murrine_draw.c
+++ b/src/murrine_draw.c
@@ -1625,10 +1625,6 @@ murrine_draw_scrollbar_slider (cairo_t *cr,
	}

	cairo_restore (cr);
-
-	murrine_set_color_rgb (cr, &border);
-	murrine_rounded_rectangle (cr, 0.5, 0.5, width-1, height-1, widget->roundness, widget->corners);
-	cairo_stroke (cr);
 }

 static void
@@ -1791,21 +1787,15 @@ murrine_draw_normal_arrow (cairo_t *cr,
 {
	double arrow_width;
	double arrow_height;
-	double line_width_2;

-	cairo_save (cr);
-
-	arrow_width = MIN (height*2.0 + MAX (1.0, ceil (height*2.0/6.0*2.0)/2.0)/2.0, width);
-	line_width_2 = MAX (1.0, ceil (arrow_width/6.0*2.0)/2.0)/2.0;
-	arrow_height = arrow_width/2.0+line_width_2;
+	arrow_width = CLAMP (width, 3, 8);
+	arrow_height = CLAMP (height, 3, 8);

-	cairo_translate (cr, x, y-arrow_height/2.0);
+	cairo_save (cr);

-	cairo_move_to (cr, -arrow_width/2.0, line_width_2);
-	cairo_line_to (cr, -arrow_width/2.0 + line_width_2, 0);
-	cairo_arc_negative (cr, 0, arrow_height-2*line_width_2-2*line_width_2*sqrt(2), 2*line_width_2, M_PI_2+M_PI_4, M_PI_4);
-	cairo_line_to (cr, arrow_width/2.0-line_width_2, 0);
-	cairo_line_to (cr, arrow_width/2.0, line_width_2);
+	cairo_translate (cr, x, y - arrow_height / 2.0);
+	cairo_move_to (cr, -arrow_width / 2.0, 0);
+	cairo_line_to (cr, arrow_width / 2.0, 0);
	cairo_line_to (cr, 0, arrow_height);
	cairo_close_path (cr);
