diff --git a/modules/basic/.libs/basic-coretext.o b/modules/basic/.libs/basic-coretext.o
index f013cbc..270a4f4 100644
Binary files a/modules/basic/.libs/basic-coretext.o and b/modules/basic/.libs/basic-coretext.o differ
diff --git a/modules/basic/.libs/pango-basic-coretext.so b/modules/basic/.libs/pango-basic-coretext.so
index 2cbf3ef..fec3e0f 100755
Binary files a/modules/basic/.libs/pango-basic-coretext.so and b/modules/basic/.libs/pango-basic-coretext.so differ
diff --git a/modules/basic/basic-coretext.c b/modules/basic/basic-coretext.c
index 82c2b48..c34460a 100644
--- a/modules/basic/basic-coretext.c
+++ b/modules/basic/basic-coretext.c
@@ -219,13 +219,17 @@ run_iterator_create (struct RunIterator *iter,
   char *copy;
   CFDictionaryRef attributes;
   CFAttributedStringRef attstr;
+  int val = 0;
+  CFNumberRef number = CFNumberCreate (kCFAllocatorDefault, kCFNumberIntType, &val);

   CFTypeRef keys[] = {
-      (CFTypeRef) kCTFontAttributeName
+      (CFTypeRef) kCTFontAttributeName,
+      kCTLigatureAttributeName
   };

   CFTypeRef values[] = {
-      ctfont
+      ctfont,
+      number
   };

   /* Initialize RunIterator structure */
@@ -238,7 +242,7 @@ run_iterator_create (struct RunIterator *iter,
   attributes = CFDictionaryCreate (kCFAllocatorDefault,
                                    (const void **)keys,
                                    (const void **)values,
-                                   1,
+                                   sizeof (keys) / sizeof (keys[0]),
                                    &kCFCopyStringDictionaryKeyCallBacks,
                                    &kCFTypeDictionaryValueCallBacks);

@@ -262,6 +266,7 @@ run_iterator_create (struct RunIterator *iter,
   iter->line = CTLineCreateWithAttributedString (attstr);
   iter->runs = CTLineGetGlyphRuns (iter->line);

+  CFRelease (number);
   CFRelease (attstr);
   CFRelease (attributes);
